### PR TITLE
Refactor plant inversion feedforward, enhancements to docs, and add LTV Unicycle controller tests.

### PR DESCRIPTION
**Pull request by cttdev**
Original PR: #17

### CODE REVIEW

**Comment by mcm001** [comment]
@calcmogul

File: wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h [hunk 1]
```diff
@@ -151,24 +135,17 @@ class PlantInversionFeedforward {
   }
 
   /**
-   * Calculate the feedforward with current anf future reference vectors.
+   * Calculate the feedforward with current and future reference vectors.
    *
-   * @param r     The current reference state of time k.
-   * @param nextR The future reference state of time k + dt.
+   * @param r     The reference state of the current timestep(k).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h]: There should be a space before the parentheses.

File: wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h [hunk 2]
```diff
@@ -132,16 +103,29 @@ class PlantInversionFeedforward {
...
+   * Calculate the feedforward with only the desired
+   * future reference. This uses the internally stored "current"
+   * reference.
    *
-   * @param nextR The future reference state of time k + dt.
+   * If this method is used the initial state of the system is the one
+   * set using Reset(const Eigen::Matrix<double, States, 1>&).
+   * If the initial state is not set it defaults to a zero vector.
+   *
+   * @param nextR The reference state of the future timestep(k + dt).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h]: There should be a space before the parentheses.

File: wpilibj/src/main/java/edu/wpi/first/wpilibj/controller/ControlAffinePlantInversionFeedforward.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  /**
+   * Calculate the feedforward with only the desired
+   * future reference. This uses the internally stored "current"
+   * reference.
+   *
+   * <p>If this method is used the initial state of the system is the one
+   * set using {@link LinearPlantInversionFeedforward#reset(Matrix)}.
+   * If the initial state is not set it defaults to a zero vector.
+   *
+   * @param nextR The reference state of the future timestep(k + dt).
```
Comment (calcmogul) [inline wpilibj/src/main/java/edu/wpi/first/wpilibj/controller/ControlAffinePlantInversionFeedforward.java]: There should be a space before the parentheses.

File: wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+  /**
+   * Calculate the feedforward with only the desired
+   * future reference. This uses the internally stored "current"
+   * reference.
+   *
+   * If this method is used the initial state of the system is the one
+   * set using Reset(const Eigen::Matrix<double, States, 1>&).
+   * If the initial state is not set it defaults to a zero vector.
+   *
+   * @param nextR The reference state of the future timestep(k + dt).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h]: There should be a space before the parentheses.

File: wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+   */
+  Eigen::Matrix<double, Inputs, 1> Calculate(
+      const Eigen::Matrix<double, States, 1>& nextR) {
+    return Calculate(m_r, nextR);
+  }
+
+  /**
+   * Calculate the feedforward with current and future reference vectors.
+   *
+   * @param r     The reference state of the current timestep(k).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h]: There should be a space before the parentheses.

File: wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+   * @param nextR The reference state of the future timestep(k + dt).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/ControlAffinePlantInversionFeedforward.h]: There should be a space before the parentheses.

File: wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h [hunk 7]
```diff
@@ -151,24 +135,17 @@ class PlantInversionFeedforward {
   }
 
   /**
-   * Calculate the feedforward with current anf future reference vectors.
+   * Calculate the feedforward with current and future reference vectors.
    *
-   * @param r     The current reference state of time k.
-   * @param nextR The future reference state of time k + dt.
+   * @param r     The reference state of the current timestep(k).
+   * @param nextR The reference state of the future timestep(k + dt).
```
Comment (calcmogul) [inline wpilibc/src/main/native/include/frc/controller/LinearPlantInversionFeedforward.h]: There should be a space before the parentheses.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by mcm001** [comment]
Should be merged

**Pull request merged**